### PR TITLE
Cygwin hooks for stdenvNative and cygwin fixes for various packages

### PR DESCRIPTION
**Pull request by chaoflow**
Todo:
- [x] rebased onto current master
- [x] publish [technical report](http://ternaris.com/lab/nix-on-windows.html) detailing this work
- [x] Verify perl 5.20 builds with windows, it was perl 5.16 when we tested the last time.
- [x] proof-read rebased diff that non-cygwin stdenv is untouched
- [x] travis error(s) - I don't understand the current error

### CODE REVIEW

**Comment by peti** [comment]
Could you please rebase this branch relative to the current `master` and force-push the new version?

**Comment by garbas** [comment]
@peti i'll do it during next week

**Comment by lethalman** [comment]
So much to review. Maybe this can be splitted in multiple commits? Don't know.

File: pkgs/applications/networking/browsers/w3m/cygwin.patch [hunk 1]
```diff
@@ -0,0 +1,1710 @@
+diff -ur w3m-0.5.3/config.guess new/w3m-0.5.3/config.guess
```
Comment (copumpkin) [inline pkgs/applications/networking/browsers/w3m/cygwin.patch]: Was this just generated by autoconf? Might be simpler (and easier to review) to just call autoconf as part of the build, if that doesn't break stuff

**Comment by copumpkin** [comment]
Some of the .patch files in there are huge. Unless you made all the changes yourself (in which case we should probably look at them _very_ carefully), It seems like it would be cleaner to just refer to wherever the new versions of the source files come from.

**Comment by chaoflow** [comment]
@copumpkin: The patches are mostly taken from cygwin tarballs, i.e. there are no direct urls to them.

File: pkgs/tools/typesetting/asciidoc/default.nix [hunk 2]
```diff
@@ -255,6 +255,7 @@ stdenv.mkDerivation rec {
   '';
 
   preInstall = "mkdir -p $out/etc/vim";
+  makeFlags = if stdenv.isCygwin then "DESTDIR=/." else null;
```
Comment (chaoflow) [inline pkgs/tools/typesetting/asciidoc/default.nix]: It's only asciidoc. Without this "fix" it produces a "//" at the beginning of the path which on cygwin is translated to "\\" which is the beginning of network SAMBA path.

**Comment by garbas** [comment]
@chaoflow @copumpkin what if we create derivation for those patches and extract patches ? it should be possible to refer to patches easier to keep them up to date.

File: pkgs/applications/version-management/git-and-tools/default.nix [hunk 3]
```diff
@@ -8,7 +8,7 @@ let
   gitBase = lib.makeOverridable (import ./git) {
     inherit fetchurl stdenv curl openssl zlib expat perl python gettext gnugrep
       asciidoc xmlto docbook2x docbook_xsl docbook_xml_dtd_45 libxslt cpio tcl
-      tk makeWrapper subversionClient gzip;
+      tk makeWrapper subversionClient gzip libiconvOrNull;
```
Comment (edolstra) [inline pkgs/applications/version-management/git-and-tools/default.nix]: Note that `libiconvOrNull` will probably go away soon: https://github.com/NixOS/nixpkgs/pull/6166

Comment (chaoflow) [inline pkgs/applications/version-management/git-and-tools/default.nix]: I understand #6199 in the sense that libiconv will replace the functionality of libiconvOrNull. Either this is merged first and libiconvOrNull occurrences will be fixed as part of #6199 or we do it the other way round.

**Comment by vcunat** [comment]
Apart from the "sed-i" issue, non-cygwin impact seems relatively safe, but it's still a mass rebuild, and it will be better to screen this for larger build regressions before merging to staging.

As for the Travis error, it sounds like an escaping issue unrelated to this. We have ``python27Packages."plone.synchronize-1.0.1"``, so there are dots *inside* the attribute name (which is really ugly). The message seems like the quotation is missed (searching for ``plone`` attribute).

**Comment by vcunat** [comment]
Patches: as suggested, on your place I'd consider creating a derivation that extracts all the cygwin patches, so it's easier to update (once upon a time); then they could be used like ``optional stdenv.isCygwin "${cygwinPatches}/pkgconfig-foo.patch"``.

**Comment by copumpkin** [comment]
Is this still alive? It seems exciting!

**Comment by garbas** [comment]
@copumpkin still alive. i hope to find time to finish this soon.

**Comment by garbas** [comment]
"if" was used instead of "stdenv.lib.optional/stdenv.lib.optionals/stdenv.lib.optionalString" to not trigger stdenv rebuild.

**Comment by vcunat** [comment]
I see, but for the sake of readability... on staging we do a stdenv rebuild almost every week.

**Comment by garbas** [comment]
@vcunat this PR is being here for almost half a year. it is really hard to keep it up to date because of patches changing for every version. i would really like to merged it as is, otherwise we would lose what we have currently working.

**Comment by vcunat** [comment]
OK, I thought the PR is waiting for you to finish something
> i hope to find time to finish this soon.

**Comment by garbas** [comment]
it is an old comment @chaoflow and I rebased it 2 days ago.

**Comment by garbas** [comment]
manually merged [here](http://git.io/vkGx4)

**Comment by vcunat** [comment]
I see; github doesn't notify participants when you add or change PR commits.

**Comment by chaoflow** [comment]
Meanwhile, we got around to publish the report: http://ternaris.com/lab/nix-on-windows.html

**Pull request closed without merging**